novobi_henry_book/models/library_booklocation.py

```diff
@@ -0,0 +1,33 @@
+from odoo import models, fields, api, exceptions
+
+
+class LibraryBookLocation(models.Model):
+    _name = 'library.booklocation'
+    _description = 'Library Book Location'
+
+    name = fields.Char('Name')
+    book_ids = fields.One2many('library.book', 'id', string = "Books", compute = '_compute_total_book')
+    total_available_books = fields.Integer('Total Available Book', compute='_compute_total_book')
+    total_books = fields.Integer('Total Book', compute='_compute_total_book')
+
+
+    # Rule: do not use sql constraint
+    # _sql_constraints = [
+    # ('name', 'unique(name)', 'Duplicate Book Location Name')
+    # ]
+
+    @api.constrains('name')
+    def _check_name(self):
+        self.ensure_one()
+        # loc_counts = self.search_count([('name', '=', self.name)])
+        loc_counts = self.env['library.booklocation'].search([('name', '=', self.name), ('id', '!=', self.id)])
+        print(loc_counts)
+        if loc_counts:
+            raise exceptions.ValidationError("Location already exists!")
+
+    # @api.depends('book_ids')
+    def _compute_total_book(self):
+        for record in self:
+            record.total_available_books = record.env['library.book'].search_count(['&',('location_id', '=', record.name),('status', '=', 'available')])
+            record.total_books = record.env['library.book'].search_count([('location_id', '=', record.name)])
+            record.book_ids = record.env['library.book'].search([('location_id', '=', record.name)])
```

Comment:
we should declare something like `LibraryBook = self.env['library.book']` before entering the `for` loop to avoid using `record.env['library.book']` too many times